### ISSUE TITLE
Matrix Multiplication Fails: InvalidArgumentError (see above for traceback): assertion failed: [] [Condition x == y did not hold element-wise:]

### ISSUE THREAD

**Issue by laurendelong21** [babylonhealth/rgat#4]
I would like to use the RGAT on some biomedical data, but I am running into some problems, and I was hoping you could give me insight as to how to troubleshoot this.

I modelled my code off of the RDF use-case which is provided, and even went so far as to try to integrate my data directly onto a forked version of your original RGAT repo, but I continue to run into the same error. The error, specifically, is:

InvalidArgumentError (see above for traceback): assertion failed: [] [Condition x == y did not hold element-wise:] [x (rgatn_model/rgat/batched_sparse_dense_matmul/strided_slice:0) = ] [12992] [y (rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_shapes/y:0) = ] [1600]

The traceback is included at the bottom of this issue.

It happens specifically in the relational_graph_attention part of the rgat code where the output layer is computed by multiplying a sparse tensor with a dense tensor. I understand that the error is thrown because the two matrices do not have compatible dimensions for matmult, but I am struggling to understand why my input data causes this to happen while the original RDF input datasets do not throw the error.

Naturally, one might think that my input data types are incorrect, but I compared my input data types against those of the AIFB dataset, and I can not find anything that seems to stand out as different. Specifically, my features are encoded as a sparse tensor, my supports are an OrderedDict with keys as relations, and values being the sparse tensors representing the respective adjacency matrices, and my get_splits function return indices as 1D arrays and labels as 2D arrays in which each row represents the labels of a node.

´  File "adrgat/cli.py", line 107, in main
    eval_input_fn, steps=1, throttle_secs=0))
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/estimator/training.py", line 471, in train_and_evaluate
    return executor.run()
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/estimator/training.py", line 610, in run
    return self.run_local()
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/estimator/training.py", line 711, in run_local
    saving_listeners=saving_listeners)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 354, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1207, in _train_model
    return self._train_model_default(input_fn, hooks, saving_listeners)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1237, in _train_model_default
    features, labels, model_fn_lib.ModeKeys.TRAIN, self.config)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1195, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "/Users/ldelong/adr-rgat/adrgat/model.py", line 98, in model_fn
    logits = model(inputs=inputs, support=support)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/keras/engine/base_layer.py", line 757, in __call__
    outputs = self.call(inputs, *args, **kwargs)
  File "/Users/ldelong/adr-rgat/adrgat/model.py", line 38, in call
    x = self.rgat1(inputs=inputs, support=support)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/keras/engine/base_layer.py", line 757, in __call__
    outputs = self.call(inputs, *args, **kwargs)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/rgat/layers/relational_graph_attention.py", line 313, in call
    sparse_tensor=attn_h_n_rm, dense_tensor=outputs_h_rm_u)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/rgat/ops/math_ops.py", line 102, in batched_sparse_dense_matmul
    sparse_shape[2], dense_shape[1], name="assert_compatible_shapes")
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/ops/check_ops.py", line 390, in assert_equal
    return control_flow_ops.Assert(condition, data, summarize=summarize)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/util/tf_should_use.py", line 189, in wrapped
    return _add_should_use_warning(fn(*args, **kwargs))
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/ops/control_flow_ops.py", line 159, in Assert
    return gen_logging_ops._assert(condition, data, summarize, name="Assert")
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/ops/gen_logging_ops.py", line 52, in _assert
    name=name)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): assertion failed: [] [Condition x == y did not hold element-wise:] [x (rgatn_model/rgat/batched_sparse_dense_matmul/strided_slice:0) = ] [12992] [y (rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_shapes/y:0) = ] [1600]
         [[node rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_shapes/Assert/Assert (defined at /Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/rgat/ops/math_ops.py:102)  = Assert[T=[DT_STRING, DT_STRING, DT_STRING, DT_INT32, DT_STRING, DT_INT32], summarize=3, _device="/job:localhost/replica:0/task:0/device:CPU:0"](rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_shapes/Equal, rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_batch_sizes/Assert/Assert/data_0, rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_batch_sizes/Assert/Assert/data_1, rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_shapes/Assert/Assert/data_2, rgatn_model/rgat/batched_sparse_dense_matmul/strided_slice, rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_shapes/Assert/Assert/data_4, rgatn_model/rgat/batched_sparse_dense_matmul/assert_compatible_shapes/y)]]
´

**Comment by franciscovargas** [comment]
@danesherbs Maybe you can help out here ?  Sorry for the late response I think most people that have used / worked on this package are no longer supporting it (moved institutions).

@laurendelong21 could you share the snippet that causes this error ? maybe you can reproduce the error by creating a randomly initialised matrix as your data instead with the same shape/type as your data, then it should be easier to debug for one of the co-authors. I have a guess at what the issue is and it might be in your setup (should be something simple like a tranpose or providing the wrong arguments for relations, units ect).

If its not too much to ask could you maybe print the shapes attn_h_n_rm and outputs_h_rm_u of  in this line :

https://github.com/babylonhealth/rgat/blob/224bae9b682f47fb9adbce9b09e3f61804caafc7/rgat/layers/relational_graph_attention.py#L313

this is not difficult to do, just open your editor with either sudo or admin access (on windows) and open to the file /Users/ldelong/adr-rgat/rgat-36/lib/python3.6/site-packages/rgat/layers/relational_graph_attention.py  and edit accordingly. 

The dimensions third dimension of attn_h_n_rm should be self._nodes*self._relations and if you look at where the assertion is called in https://github.com/babylonhealth/rgat/blob/224bae9b682f47fb9adbce9b09e3f61804caafc7/rgat/ops/math_ops.py#L102  its on the third dimension. 

This means that the number of nodes (or relations) in your data (dense_tensor ) is not the same as the number of nodes obtained here https://github.com/babylonhealth/rgat/blob/224bae9b682f47fb9adbce9b09e3f61804caafc7/rgat/layers/relational_graph_attention.py#L265 differs to that . 

So my guess is that the inputs you are providing when calling the  RGAT object (inputs argument https://github.com/babylonhealth/rgat/blob/224bae9b682f47fb9adbce9b09e3f61804caafc7/rgat/layers/relational_graph_attention.py#L253)  has the number of nodes in the wrong dimension so transposing your data may fix this. 

To reiterate the dimension of inputs seems to be (number of nodes, relations * units * heads). The other possible mistake you could have made is to pass in the wrong number of relations as when initialising the object relative to the second dimension of the inputs. See https://github.com/babylonhealth/rgat/blob/224bae9b682f47fb9adbce9b09e3f61804caafc7/examples/batching/example_static.py#L55

**Comment by laurendelong21** [comment]
Thanks a lot for your response. Transposing my input matrix worked!

**Comment by laurendelong21** [comment]
BTW- according to the [usage guide](https://github.com/babylonhealth/rgat/blob/master/docs/usage_guide.md) it seems that the input matrix should be the other way around, which is why I made it like that originally.

**Comment by franciscovargas** [comment]
If you look at the user guide there is this line:

> inputs = get_inputs()    # Dense tensor with shape (?, Features)

this tells you inputs has 2 dimensions the first one is associated with "?" this typically refers to the size of the dataset / number of data points which in the case of a graph is the number of nodes, so I think it is conveying (number_of_nodes, number_of_features) as the input shape which is correct. I do agree its not the most clearly written blurb but I think its correct. Again deferring to one of the authors @danesherbs to double check this.

**Comment by danesherbs** [comment]
Thanks @franciscovargas! That's correct - the inputs should have shape `(Nodes, Features)`.
